pkg/flow/flow_mem_driver.go

```diff
@@ -810,25 +810,22 @@ func (fc *FlowCollector) updateRecord(record interface{}) error {
 			} else {
 				if link.EndTime > 0 {
 					if current.Direction != nil {
-						var routerId string
+						var peerRouterId string
 						peerDirection := Outgoing
 						if *current.Direction == Outgoing {
 							peerDirection = Incoming
 						}
-						// link being deleted, find peer router entry
+						// find any links to the peer router and delete the one that is for current link
 						for _, router := range fc.Routers {
 							if *current.Name == normalizeRouterName(*router.Name) {
-								routerId = router.Identity
+								peerRouterId = router.Identity
 								break
 							}
 						}
-
-						// find any links to the peer router and delete
 						for _, eLink := range fc.Links {
-							if eLink.Direction != nil && *eLink.Direction == peerDirection {
-								if routerId == eLink.Parent {
-									fc.deleteRecord(eLink)
-								}
+							if eLink.Direction != nil && *eLink.Direction == peerDirection && peerRouterId == eLink.Parent && *eLink.Name == normalizeRouterName(*fc.Routers[current.Parent].Name) {
```

Comment:
Logic looks solid to me!

Still a bit afraid of a panic if eLink.Name is nil, if fc.Routers[current.Parent] is not present (nil) or if fc.Routers[current.Parent].Name is nil. All three of those should be easy to contrive in the test we have (even if we don't actually commit those test cases.) I am not so sure how likely it'd be to occur in even the most troubled real networks, but I'd rather not be surprised.